File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        maxLen = 0
        track = set()
        left = 0
        for right in range(len(s)):
            while s[right] in track:
                track.remove(s[left])
                left += 1
            track.add(s[right])
            maxLen = max(right - left + 1, maxLen)
        return maxLen

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
import java.util.*;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> seen = new HashSet<>();
        int left = 0, maxLen = 0;
        for(int right = 0;right < s.length();right++){
            while(seen.contains(s.charAt(right))){
                seen.remove(s.charAt(left));
                left++;
            }
            seen.add(s.charAt(right));
            maxLen = Math.max(right-left+1, maxLen);
        }
        return maxLen;
    }
}